server/iframe.go

```diff
@@ -46,30 +53,175 @@ func (a *API) iFrame(w http.ResponseWriter, _ *http.Request) {
+		return
+	}
+
+	config := a.p.apiClient.Configuration.GetConfig()
+
+	enableDeveloper := config.ServiceSettings.EnableDeveloper
+
+	// Ideally we'd accept the token via an Authorization header, but for now get it from the query string.
+	// token := r.Header.Get("Authorization")
+	token := r.URL.Query().Get("token")
+
+	// Validate the token in the request, handling all errors if invalid.
+	expectedTenantIDs := []string{a.p.getConfiguration().TenantID}
+	claims, validationErr := validateToken(a.p.tabAppJWTKeyFunc, token, expectedTenantIDs, enableDeveloper != nil && *enableDeveloper)
+	if validationErr != nil {
+		handleErrorWithCode(logger, w, validationErr.StatusCode, validationErr.Message, validationErr.Err)
+		return
+	}
+
+	oid, ok := claims["oid"].(string)
+	if !ok {
+		logger.Error("No claim for oid")
+		http.Error(w, "", http.StatusBadRequest)
+		return
+	}
+
+	logger = logger.WithField("oid", oid)
+
+	ssoUsername, ok := claims["unique_name"].(string)
+	if !ok {
+		logger.Warn("no unique_name claim")
+
+		ssoUsername, ok = claims["preferred_username"].(string)
+		if !ok {
+			logger.Error("No claim for unique_name or preferred_username")
+			http.Error(w, "", http.StatusBadRequest)
+			return
+		}
+	}
+
+	mmUser, err := a.p.apiClient.User.GetByEmail(ssoUsername)
+	if err != nil && err != pluginapi.ErrNotFound {
+		logger.WithError(err).Error("Failed to query Mattermost user matching unique_name")
+		http.Error(w, "Internal server error", http.StatusInternalServerError)
+		return
+	} else if mmUser == nil {
+		logger.Warn("No Mattermost user matching unique_name, redirecting to login")
+
+		// Redirect to the home page
+		http.Redirect(w, r, "/", http.StatusSeeOther)
+		return
+	}
+
+	logger = logger.WithField("user_id", mmUser.Id)
+
+	if mmUser.DeleteAt != 0 {
+		logger.Warn("Mattermost user is archived, redirecting to login")
+
+		// Redirect to the home page
+		http.Redirect(w, r, "/", http.StatusSeeOther)
+		return
+	}
+
+	// Keep track of the unique_name and oid in the user's properties to support
+	// notifications in the future.
+	mmUser.Props["com.mattermost.plugin-msteams-devsecops.sso_username"] = ssoUsername
+	mmUser.Props["com.mattermost.plugin-msteams-devsecops.oid"] = oid
+
+	err = a.p.apiClient.User.Update(mmUser)
+	if err != nil {
+		logger.WithError(err).Error("Failed to update Mattermost user with claims")
+		http.Error(w, "Internal server error", http.StatusInternalServerError)
+		return
+	}
+
+	// This is effectively copied from https://github.com/mattermost/mattermost/blob/a184e5677d28433495b0cde764bfd99700838740/server/channels/app/login.go#L287
+	secure := true
+	maxAgeSeconds := *config.ServiceSettings.SessionLengthWebInHours * 60 * 60
+	domain := getCookieDomain(config)
+	subpath, _ := utils.GetSubpathFromConfig(config)
+
+	expiresAt := time.Unix(model.GetMillis()/1000+int64(maxAgeSeconds), 0)
+
+	session, err := a.p.apiClient.Session.Create(&model.Session{
+		UserId:   mmUser.Id,
+		DeviceId: model.NewId(),
```

Comment:
`Session.DeviceId` is used to indicate the session is on a mobile device.  Why do we want to set that in this case?

Reply:
Ooh, that's a good catch! No, we shouldn't do that.